### ISSUE TITLE
Still no alternative to onCompleted change in 3.8

### ISSUE THREAD

**Issue by flashtheman** [apollographql/apollo-client#12056]
### Issue Description

In Apollo Client 3.8, the behavior changed such that onCompleted is no longer called when another component triggers a refetch of the corresponding query. In versions <= 3.7.x, onCompleted was always invoked when notifyOnNetworkStatusChange was set to true.

We have now been waiting for a year for an alternative to the old behavior, as our app heavily relies on this functionality.

Is there any plan to provide an alternative, or is there an established workaround? The only solution we know of is using useEffect to watch the data object, but this would require significant changes to our application. Because of that, we've been hoping for an official solution to this issue.

### Link to Reproduction

none

### `@apollo/client` version

3.8.x

**Comment by jerelmiller** [comment]
Hey @flashtheman 👋 

To clarify, are you saying that after upgrading to > 3.8, `notifyOnNetworkStatusChange: true` is not triggering the `onCompleted` callback? If so, that is a bug. A runnable reproduction would be helpful to understand what's happening here if you can provide one.

---

As for the bad news. At this time we aren't planning to introduce anymore changes to `onCompleted` or introduce an alternative callback in its place. That change in 3.8 ended up being way more spicy than we had anticipated and is one of those changes we wished we would have waited for a major version to introduce. This is unfortunately one of those APIs that is impossible to make everyone happy. Everyone has a different idea of how it should work so while we fixed the behavior in 3.8 for one camp of individuals, others used it as a feature and saw the change as an introduction of a bug.

Would you be able to provide some context for how you're using `onCompleted`? Perhaps we can give some recommendations on good/bad practices.

I'm sorry I don't have a better answer for you at this time otherwise.

**Comment by jerelmiller** [comment]
That helps! Appreciate the context!

This is one of those cases where it sort of worked by accident in 3.7.x or before. Using that object syntax with `refetchQueries` actually creates an [entirely new `ObservableQuery` instance](https://github.com/apollographql/apollo-client/blob/aee0f4fd3ef4291b9ef87ab233c8c04be26397ce/src/core/QueryManager.ts#L912-L919) under the hood (the thing returned by `client.watchQuery(...)` which `useQuery` calls [under the hood](https://github.com/apollographql/apollo-client/blob/aee0f4fd3ef4291b9ef87ab233c8c04be26397ce/src/react/hooks/useQuery.ts#L191-L193)), so these queries weren't actually connected in any meaningful way. The `onCompleted` callback was triggered purely because the refetch from the other `ObservableQuery` instance caused a cache update. As you know, cache updates in 3.7 call `onCompleted`.

The `notifyOnNetworkStatusChange` option to `refetchQueries` here doesn't actually do much for you because you don't actually have access to the `ObservableQuery` instance that was created using this syntax. I can almost guarantee that removing that `notifyOnNetworkStatusChange` option using 3.7 or less won't change the behavior you're seeing/expecting with `onCompleted`. So with that, don't worry about a runnable reproduction since this isn't actually a bug.

That said, I'm curious, what kind of logic are you executing in `onCompleted`? Perhaps I can provide some help with that logic?

**Comment by jerelmiller** [comment]
Is the state variable you're setting based on the data returned from the query? 

If so, we'd recommend using a derived value instead:

```ts
// Don't do this:
const [count, setCount] = useState(0);

const { data } = useQuery(query, {
  onCompleted: (data) => setCount(data.items.length)
});

// instead do this:
const { data } = useQuery(query);

const count = data.items.length;
```

I don't know if this is how you're using it, but if so, use a derived value which will be much more accurate since it will be kept in sync with `data` as cache writes happen.

**Comment by emadabdulrahim** [comment]
@jerelmiller Hi, where can I read about the changes made to onCompleted in 3.8? Docs have barely any mention of onCompleted.

For example, does onCompleted fire when the query get value from cache? if not, where is this information in the docs?

**Comment by jerelmiller** [comment]
@emadabdulrahim apologies, we've had an issue opened for a while to update this (#11306) but haven't found the time to get it done.

The tl;dr; with the changes in 3.8:
* `onCompleted` doesn't fire on cache updates like it does with 3.7
* `onCompleted` doesn't fire on network requests unless `notifyOnNetworkStatusChange` is set to `true`

The work was done in https://github.com/apollographql/apollo-client/pull/10229 if you'd like more context.

**Comment by emadabdulrahim** [comment]
@jerelmiller Do the changes to onCompleted also apply to useLazyQuery? or only useQuery?

> onCompleted doesn't fire on network requests unless notifyOnNetworkStatusChange is set to true

I can't seem to verify this part. Queries  where `notifyOnNetworkStatusChange` is set to false also trigger `onCompleted` anyway. Could you clarify this part?

**Comment by jerelmiller** [comment]
@emadabdulrahim they would apply to `useLazyQuery` as well since `useLazyQuery` uses `useQuery` under the hood.

> I can't seem to verify this part. Queries where `notifyOnNetworkStatusChange` is set to `false` also trigger `onCompleted` anyway. Could you clarify this?

It should still fire the _first_ time the network request is returned, but anything after that won't trigger it (`refetch`, cache update, etc.). Are you still seeing it called after the first time?

**Comment by emadabdulrahim** [comment]
No you're right. `onCompleted` only fires on subsequent requests if `notifyOnNetworkStatusChange` is set to `true`. Thanks for the speedy response!

**Comment by emadabdulrahim** [comment]
One last thing, is there anything to watch out for regarding the timing of when `onCompleted` is fired in React 17 vs 18? Would there be any regression or changes in behavior bumping to React 18 with regards to onCompleted callbacks, say for the queries where I set `notifyOnNetworkStatusChange` to `true`?

**Comment by jerelmiller** [comment]
Perhaps. May I ask what you're using `onCompleted` for?

Generally the timing shouldn't matter for anything you're doing in `onCompleted` and if does, that may indicate a code smell. I may be able to give a couple pointers depending on what you're using it for.

**Comment by emadabdulrahim** [comment]
I agree many (if not all) of the usages of onCompleted I see should be refactored/eliminated. But putting that aside, the uses can be:
- Set some state based on query data received. This could be to set some preferable default to the user. E.g. if something is available, then show this part of the UI or default to it being open. 
- Trigger some other query (waterfall) say to validate some data
- Since apollo doesn't do stale while revalidate, sometimes query data is kept around in a ref to provide certain UI experience. (though this doesn't need `onCompleted` and is easiest to refactor)


It'd say majority of uses are to set state. Those are ideally refactored to be derived from query data, and/or access from apollo cache directly instead of replicating cache in a React state or context.

**Comment by jerelmiller** [comment]
> It'd say majority of uses are to set state. Those are ideally refactored to be derived from query data, and/or access from apollo cache directly instead of replicating cache in a React state or context.

Yep I was going to say the same thing. Deriving data is your best bet and ensures you'll always be up-to-date with the latest value. You might check those anyways since cache updates don't fire `onCompleted` which means some of those `setStates` may contain stale data that isn't up-to-date with the `data` property returned from either of those query hooks.

Sounds like you have a handle on this though! I'll see if I can find some time to add a warning somewhere in our docs about setting state in `onCompleted`. Unfortunately I've seen this all-too-often 😞

**Comment by emadabdulrahim** [comment]
sorry to hammer on this, but to be absolutely certain, when you say "since cache updates don't fire `onCompleted`" this is unless `notifyOnNetworkStatusChange` is set to `true`, right?

**Comment by jerelmiller** [comment]
Not quite. By "cache updates", I mean a value written to the cache not caused by a network request in the hook, such as a cache update from a mutation or a manual cache update that you perform by some other means. `useQuery`/`useLazyQuery` compares previous/current `networkStatus` (which only changes when `notifyOnNetworkStatusChange` is `true) to determine whether to call `onCompleted` which means a network request for that query has to happen.

This was the big change in 3.8 which caused the stir since versions ~v3.4.17 to 3.8.0 would call `onCompleted` on those cache updates.

Because of this behavior, the `data` property may get updated with a new value based on that cache update, but because `onCompleted` is not fired, your state values will now be out-of-sync and you'll be rendering stale data. Does that make sense?

**Comment by emadabdulrahim** [comment]
I see. Luckily we don't have a lot of direct cache manipulation after mutations. We mainly rely on `refetch` property on useMutation hook to revalidate queries. 

Thanks for clarifying that!
